File: Algorithm_LevelOne/Permutation_in_String/Permutation_in_String.cc
// 567. Permutation in String C++
#include <string>
#include <unordered_map>

class Solution {
public:
    bool checkInclusion(std::string s1, std::string s2) {
        std::unordered_map<char, int> mp1, mp2;
        int size = s1.size();
        for (auto x: s1) mp1[x]++;
        int start = 0;
        int max = s2.size();
        for (int i = 0; i < max; i++) {
            mp2[s2[i]]++;
            if (i - start + 1 == size) {
                if (mp1 == mp2) return true;
                if (mp2[s2[start]] == 1) {
                    mp2.erase(s2[start]);
                } else {
                    mp2[s2[start]] -= 1;
                }
                start++;
            }
        }
        return false;
    }
};

File: Algorithm_LevelOne/Permutation_in_String/Permutation_in_String.py
# 567. Permutation in String Python

class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        if len(s2) < len(s1):
            return False
        s1Counter = {}
        for c in s1:
            if c not in s1Counter:
                s1Counter[c] = 0
            s1Counter[c] += 1
        
        start = 0
        s2Counter = {}
        for end, char in enumerate(s2):
            if char not in s2Counter:
                s2Counter[char] = 0
            s2Counter[char] += 1
            size = end - start + 1
            if size == len(s1):
                if s1Counter == s2Counter:
                    return True
                if s2Counter[s2[start]] == 1:
                    del s2Counter[s2[start]]
                else:
                    s2Counter[s2[start]] -= 1
                start += 1
        return False
